dali/pipeline/executor/executor2/exec_graph.cc

```diff
@@ -0,0 +1,293 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <utility>
+#include "dali/core/cuda_event_pool.h"
+#include "dali/pipeline/executor/executor2/exec_graph.h"
+#include "dali/pipeline/executor/executor2/op_task.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/graph/op_graph2.h"
+#include "dali/pipeline/operator/batch_size_provider.h"
+
+namespace dali {
+namespace exec2 {
+
+using tasking::Task;
+using tasking::Semaphore;
+using tasking::Scheduler;
+
+inline OpType BackendFromOp(const OperatorBase *op) {
+  if (dynamic_cast<const Operator<CPUBackend>*>(op)) {
+    return OpType::CPU;
+  } else if (dynamic_cast<const Operator<GPUBackend>*>(op)) {
+    return OpType::GPU;
+  } else {
+    assert(dynamic_cast<const Operator<MixedBackend>*>(op));
+    return OpType::MIXED;
+  }
+}
+
+ExecNode::ExecNode(std::unique_ptr<OperatorBase> op, const graph::OpNode *def)
+: op(std::move(op))
+, instance_name(def ? def->instance_name : "")
+, backend(def ? def->op_type : op ? BackendFromOp(op.get()) : OpType::CPU)
+, is_batch_size_provider(dynamic_cast<BatchSizeProvider *>(this->op.get()) != nullptr) {
+  if (def) {
+    outputs.resize(def->outputs.size());
+    inputs.resize(def->inputs.size());
+  }
+  if (this->op) {
+    auto &spec = this->op->GetSpec();
+    assert(!def || def->outputs.size() == static_cast<size_t>(spec.NumOutput()));
+    assert(!def || def->inputs.size() == static_cast<size_t>(spec.NumInput()));
+    outputs.resize(std::max<size_t>(outputs.size(), spec.NumOutput()));
+    inputs.resize(std::max<size_t>(inputs.size(), spec.NumInput()));
+  }
+}
+
+void ExecNode::CreateMainTask(const WorkspaceParams &params) {
+  main_task_ = OpTask::CreateTask(this, params);
+  if (prev_task_)
+    main_task_->Succeed(prev_task_);
+}
+
+void ExecNode::CreateAuxTasks() {
+  // The concurrency of the operator may be limited within a group of operators sharing
+  // one concurrency semaphore.
+  // Note that operator cannot run parallel to its previous iterations and we add a temporal
+  // dependency between the previous and current iteration.
+  if (concurrency)
+    main_task_->GuardWith(concurrency);
+
+  // The output queue depth may be limited - this is guarded by a semaphore with initial count
+  // equal to the queue depth.
+  // The task will acquire this semaphore prior to running (potentially delaying the start).
+  // The semaphore is released when all consumers of the current task's outputs are complete.
+  // This means that nobody is accessing those outputs and they've been disposed of (unless
+  // forwarded down the pipeline, but we're ok with that).
+  if (output_queue_limit) {
+    release_outputs_ = Task::Create([]() {});
+    release_outputs_->ReleaseAfterRun(output_queue_limit);
```

Comment:
When should `Succeed` be used in case of calling `ReleaseAfterRun` as is the case with `GuardWith`? I kinda expected this as a requirement

Reply:
It's not "guarded". `main_task_` acquires the semaphore, `release_outputs_` releases it.